src/Microsoft.DotNet.Interactive.ExtensionLab/MsSqlKernel.cs

```diff
@@ -3,21 +3,219 @@
 
 using System.Data.Common;
 using Microsoft.Data.SqlClient;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.DotNet.Interactive.Commands;
+using Microsoft.DotNet.Interactive.Events;
+using System.Data;
+using System.Reactive.Disposables;
+using System.Threading;
+using Microsoft.DotNet.Interactive.Formatting;
+using Newtonsoft.Json.Linq;
+using System.Collections.Concurrent;
+using System.Text;
+using Microsoft.DotNet.Interactive.Utility;
 
 namespace Microsoft.DotNet.Interactive.ExtensionLab
 {
-    public class MsSqlKernel : SqlKernel
+    public class MsSqlKernel :
+        Kernel,
+        IKernelCommandHandler<SubmitCode>,
+        IKernelCommandHandler<RequestCompletions>,
+        IKernelCommandHandler<RequestHoverText>
     {
+        private bool _connected = false;
+        private bool _intellisenseReady = false;
+        private readonly string _tempFilePath;
+        private readonly string _tempFileUri;
         private readonly string _connectionString;
+        private readonly MsSqlServiceClient _serviceClient;
+
+        private TaskCompletionSource<ConnectionCompleteParams> _connectionCompleted = new TaskCompletionSource<ConnectionCompleteParams>();
+
+        private Func<QueryCompleteParams, Task> _queryCompletionHandler = null;
 
         public MsSqlKernel(string name, string connectionString) : base(name)
         {
+            _tempFilePath = Path.GetTempFileName();
+            _tempFileUri = MsSqlServiceClient.GetUriForFilePath(_tempFilePath);
             _connectionString = connectionString;
+
+            _serviceClient = MsSqlServiceClient.Instance;
+            _serviceClient.Initialize();
+
+            _serviceClient.OnConnectionComplete += HandleConnectionComplete;
+            _serviceClient.OnQueryComplete += HandleQueryCompleteAsync;
+            _serviceClient.OnIntellisenseReady += HandleIntellisenseReady;
+
+            RegisterForDisposal(() =>
+            {
+                if (_connected)
+                {
+                    _serviceClient.DisconnectAsync(_tempFileUri).Wait();
```

Comment:
This could probably stand to have a timeout.

To avoid the possibility of a deadlock in the presence of a `SynchronizationContext` (which you have no control over here), you can do this:

```csharp
Task.Run(() => _serviceClient.DisconnectAsync(_tempFileUri)).Wait();
```

Reply:
What's the difference here versus just waiting on the disconnect task itself?

Reply:
If there's a `SynchronizationContext` present, then the task will be scheduled on that context, at which point calling a blocking method such as `Task.Wait` or `Task.Result` can result in a deadlock. 

Calling `Task.Run` will ensure that the task is not scheduled on the current `SynchronizationContext` if there is one.